### PR TITLE
Use cached_property for much of `Ga`

### PR DESCRIPTION
**Pull request by eric-wieser**
A reduced version of #252, which will hopefully pass CI.

This seems to go with the spirit of https://github.com/pygae/galgebra/pull/78#pullrequestreview-323373088

### CODE REVIEW

**Comment by eric-wieser** [comment]
Failure is probably #257

**Comment by eric-wieser** [comment]
Unreleased sphinx version is needed to make the docs pretty, but we can survive without for now:

image

File: galgebra/_cached_property.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+
+class cached_property:
```
Comment (utensil) [inline galgebra/_cached_property.py:2]: Can this file be placed somewhere else? Don't like `_*.py` to appear at the root, they would confuse first comers if there'll be more...

P.S. Is it ported or backported from somewhere? Then `_backports` might fit? (B.T.W. I didn't like this restrictive name last time too, these small utility functions or classes are ported here to keep the dependency minimal, they deserve a dedicated place).

Comment (eric-wieser) [reply galgebra/_cached_property.py:2]: So, the purpose of `_*.py` is to indicate that the module is private, and for internal use only. If this were called `galgebra.cached_property`, then that would imply to users that they can use it themselves, and would mean that we would have to support it forever.

So I'm quite attached to using an `_` at the start of the filename, but I'd also be happy to pick a more generic filename to let us pile more things in here.

Comment (eric-wieser) [reply galgebra/_cached_property.py:2]: `_core/cached_property.py`?

Comment (utensil) [reply galgebra/_cached_property.py:1]: `galgebra/_utils/cached_property.py` 😄

Comment (eric-wieser) [reply galgebra/_cached_property.py]: Done

**Comment by utensil** [comment]
Circle CI build time is very unstable:

CI                   | before | after
---------------------|--------|--------
python-3.5           | 5m 10s | 5m 26s
python-3.6           | 7m 38s | 5m 17s
python-3.7           | 5m  9s | 5m 32s
python-3.7-symengine | 5m 18s | 8m 39s
python-3.7-sympy-1.3 | 4m 35s | 5m 3s
python-3.8           | 5m 37s | 4m 38s

**Comment by eric-wieser** [comment]
> private classes/functions are now in `_utils`, but the data structures in public API should also be grouped together somewhere else, e.g. `struct.`

Agreed. Will make a patch for that tomorrow.

> The classes in 1 should have consistent casing, i.e. `lazy_dict` should be `LazyDict`

`lazy_dict` was motivated by `dict` and `defaultdict`, but I guess that consistency isn't worth it. Will make a patch for that tomorrow.

> The properties are better ordered by their dependency depth.

My intent was to try and order in a way that makes sense in the docs, now that #254 is merged. Perhaps those orders are the same anyway.
Either way, reordering is quite destructive to history and merge conflicts, so I probably won't do that anytime soon.

> This should give a performance boost, did it?

I wouldn't expect much of a performance boost - most of these properties are needed almost immediately anyway, so the laziness isn't that useful. It's mainly a tool to avoid thoughts about initialization order, as you remarked in the other PR

**Comment by utensil** [comment]
> Either way, reordering is quite destructive to history and merge conflicts, so I probably won't do that anytime soon.

Yeah, upon reflection, ordering (of dependency depth) doesn't make much sense. It's just another way of saying that we need to be clear about the dependencies between these properties.

**Comment by utensil** [comment]
> I wouldn't expect much of a performance boost - most of these properties are needed almost immediately anyway, so the laziness isn't that useful.

I would like to verify this against https://github.com/pygae/galgebra/blob/master/examples/LaTeX/curvi_linear_latex.py#L179-L194 .

**Pull request merged**